cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java

```diff
@@ -16,46 +16,146 @@
 
 package co.cask.cdap.internal.app.preview;
 
+import co.cask.cdap.api.artifact.ArtifactScope;
 import co.cask.cdap.api.metrics.MetricTimeSeries;
 import co.cask.cdap.app.preview.PreviewManager;
 import co.cask.cdap.app.preview.PreviewStatus;
+import co.cask.cdap.app.runtime.ProgramController;
+import co.cask.cdap.app.runtime.ProgramRuntimeService;
+import co.cask.cdap.app.store.preview.PreviewStore;
 import co.cask.cdap.common.NotFoundException;
+import co.cask.cdap.internal.app.deploy.ProgramTerminator;
+import co.cask.cdap.internal.app.runtime.AbstractListener;
+import co.cask.cdap.internal.app.services.ApplicationLifecycleService;
+import co.cask.cdap.internal.app.services.ProgramLifecycleService;
+import co.cask.cdap.proto.BasicThrowable;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
 import co.cask.cdap.proto.artifact.AppRequest;
+import co.cask.cdap.proto.artifact.ArtifactSummary;
 import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.ArtifactId;
 import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramId;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.inject.Inject;
 import org.apache.twill.api.logging.LogEntry;
+import org.apache.twill.common.Threads;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import javax.annotation.Nullable;
 
 /**
  * Default implementation of the {@link PreviewManager}.
  */
 public class DefaultPreviewManager implements PreviewManager {
+  private static final Gson GSON = new Gson();
+  private static final ProgramTerminator NOOP_PROGRAM_TERMINATOR = new ProgramTerminator() {
+    @Override
+    public void stop(ProgramId programId) throws Exception {
+      // no-op
+    }
+  };
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final ProgramLifecycleService programLifecycleService;
+  private final PreviewStore previewStore;
+  private PreviewStatus status;
```

Comment:
Since request is only assigned during start and we create DefaultPreviewManager per request it should be fine. For status member variable, adding synchronized setter method.